python/pdo/test/contract.py

```diff
@@ -182,82 +245,32 @@ def CreateAndRegisterContract(config, enclave, contract_creator_keys) :
-        secret_list = secret_helper.create_secrets_for_services(
-            provisioning_services, enclave.enclave_keys, contract_id, contract_creator_id)
-
-    logger.debug('secrets: %s', secret_list)
-
-    try :
-        secretinfo = enclave.verify_secrets(contract_id, contract_creator_id, secret_list)
-        assert secretinfo
+    encrypted_state_encryption_keys = AddEnclaveSecrets(
+        ledger_config, contract.contract_id, contract_creator_keys, enclaves, provisioning_services)
 
-        encrypted_state_encryption_key = secretinfo['encrypted_state_encryption_key']
-        signature = secretinfo['signature']
+    for enclave_id in encrypted_state_encryption_keys :
+        encrypted_key = encrypted_state_encryption_keys[enclave_id]
+        contract.set_state_encryption_key(enclave_id, encrypted_key)
 
-    except Exception as e :
-        logger.error('failed to create the state encryption key; %s', str(e))
-        ErrorShutdown()
-
-    try :
-        if not secrets.verify_state_encryption_key_signature(
-                encrypted_state_encryption_key,
-                secret_list,
-                contract_id,
-                contract_creator_id,
-                signature,
-                enclave.enclave_keys) :
-            raise RuntimeError('signature verification failed')
-    except Exception as e :
-        logger.error('failed to verify the state encryption key; %s', str(e))
-        ErrorShutdown()
-
-    logger.info('encrypted state encryption key: %s', encrypted_state_encryption_key)
-
-    # --------------------------------------------------
-    logger.info('add the provisioned enclave to the contract')
-    # --------------------------------------------------
-    try :
-        if use_ledger :
-            txnid = contract_helper.add_enclave_to_contract(
-                ledger_config,
-                contract_creator_keys,
-                contract_id,
-                enclave.enclave_id,
-                secret_list,
-                encrypted_state_encryption_key,
-                signature,
-                transaction_dependency_list=txn_dependencies)
-            txn_dependencies = [ txnid ]
-
-            logger.info('contract state encryption key added to contract')
-        else :
-            logger.debug('no ledger config; skipping state encryption key registration')
-    except Exception as e :
-        logger.error('failed to add state encryption key; %s', str(e))
-        ErrorShutdown()
+    #add replication information to contract
+    AddReplicationParamsToContract(config, enclaves, contract)
 
-    contract.set_state_encryption_key(enclave.enclave_id, encrypted_state_encryption_key)
+    # Decide if the contract use a fixed enclave or a randomized one for each update. If fixed, we chose here. If random, nothing to be done here,
+    # will be (atuomatically) selected at random during request creation
+    if (use_eservice is False) or (config['Service']['Randomize_Eservice'] is False):
```

Comment:
this is a rather complicated way to get at the conditional. could invert... if use_eservice and config['Service']['Randomize_EService'] :

Reply:
sure, thank you, will fix this